crypto/merkletrie/node.go

```diff
@@ -0,0 +1,383 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package merkletrie
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+type node struct {
+	hash         []byte
```

Comment:
Can you add a comment describing what node.hash is supposed to be? Is it the full path to that node before you call `calculateHash` and then the hash-of-child-hashes afterward?

Reply:
it's a tricky one for sure. It can be one of 3 things:
as a leaf, it's the residual element.
as a non-leaf, before calling commit ( which internally calls calculateHash), it's the path to that node.
as a non-leaf, after calling commit, it's the hash of the children plus the path as implemented in calculateHash.